eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/JavadocCodeSnippetCheck.java

```diff
@@ -0,0 +1,255 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.tools.checkstyle.checks;
+
+import com.puppycrawl.tools.checkstyle.DetailNodeTreeStringPrinter;
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.DetailNode;
+import com.puppycrawl.tools.checkstyle.api.FullIdent;
+import com.puppycrawl.tools.checkstyle.api.JavadocTokenTypes;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.BlockCommentPosition;
+import com.puppycrawl.tools.checkstyle.utils.JavadocUtil;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.Locale;
+
+/**
+ * Codesnippet description should match naming pattern requirement below:
+ * <ol>
+ * <li>Package, class and method names should be concatenated by dot '.'. Ex., packageName.className.methodName</li>
+ * <li>Methods arguments should be concatenated by dash '-'. Ex. string-string  for methodName(String s, String s2)</li>
+ * <li>Use '#' to concatenate 1) and 2), ex packageName.className.methodName#string-string</li>
+ * </ol>
+ */
+public class JavadocCodeSnippetCheck extends AbstractCheck {
+
+    private static final String CODE_SNIPPET_ANNOTATION = "@codesnippet";
+    private static final String MISSING_CODESNIPPET_TAG_MESSAGE = "There is a @codesnippet block in the JavaDoc, but it does not refer to any sample.";
+
+    private static final int[] TOKENS = new int[] {
+        TokenTypes.PACKAGE_DEF,
+        TokenTypes.BLOCK_COMMENT_BEGIN,
+        TokenTypes.CLASS_DEF,
+        TokenTypes.METHOD_DEF
+    };
+
+    private String packageName;
+    // A container to contains all class name visited, remove the class name when leave the same token
+    private Deque<String> classNameStack = new ArrayDeque<>();
+    // A container to contains all METHOD_DEF node visited, remove the node whenever leave the same token
+    private Deque<DetailAST> methodDefStack = new ArrayDeque<>();
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return TOKENS;
+    }
+
+    @Override
+    public boolean isCommentNodesRequired() {
+        return true;
+    }
+
+    @Override
+    public void leaveToken(DetailAST token) {
+        switch (token.getType()) {
+            case TokenTypes.CLASS_DEF:
+                if (!classNameStack.isEmpty()) {
+                    classNameStack.pop();
+                }
+                break;
+            case TokenTypes.METHOD_DEF:
+                if (!methodDefStack.isEmpty()) {
+                    methodDefStack.pop();
```

Comment:
Why do we need Stack for this?   Would a simple String variable work?
It make sense for classes with inner classes to be able to pop() and come back to parent class...  But, can we find a new Method_DEF inside a Method_def?

Reply:
Good point. A variable should be enough.